test/e2e-library.sh

```diff
@@ -0,0 +1,154 @@
+    sleep 30
+
+    go_test_e2e -timeout=15m -failfast -parallel=1 ./test/ha -spoofinterval="10ms" \
+    --ingressClass=gateway-api.ingress.networking.knative.dev || failed=1
+
+    (( failed )) && dump_cluster_state
+    (( failed )) && fail_test
+
+    success
+}
+
+# The tests below are run in a Kind cluster, therefore they need to have their ingressendpoint and cluster-suffix specified.
+function cluster_info() {
+    CONTROL_NAMESPACE=knative-serving
+    IPS=( $(kubectl get nodes -lkubernetes.io/hostname!=kind-control-plane -ojsonpath='{.items[*].status.addresses[?(@.type=="InternalIP")].address}') )
+    CLUSTER_SUFFIX=${CLUSTER_SUFFIX:-cluster.local}
+}
+
+function kind_e2e_contour() {
+    export GATEWAY_OVERRIDE=envoy
+    export GATEWAY_NAMESPACE_OVERRIDE=contour-external
+
+    cluster_info
+
+    echo ">> Running e2e tests"
+
+    go test -race -count=1 -short -timeout=20m -tags=e2e ./test/conformance \
+    --enable-alpha --enable-beta \
+    --skip-tests="${CONTOUR_UNSUPPORTED_E2E_TESTS}" \
+    --ingressendpoint="${IPS[0]}" \
+    --ingressClass=gateway-api.ingress.networking.knative.dev \
+    --cluster-suffix="${CLUSTER_SUFFIX}"
+
+    # Give the controller time to sync with the rest of the system components.
+    sleep 30
+
+    echo ">> Scale up controller for HA tests"
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=2
+
+    go test -count=1 -timeout=15m -failfast -parallel=1 -tags=e2e ./test/ha -spoofinterval="10ms" \
+    --enable-alpha --enable-beta \
+    --ingressendpoint="${IPS[0]}" \
+    --ingressClass=gateway-api.ingress.networking.knative.dev \
+    --cluster-suffix="${CLUSTER_SUFFIX}"
+
+    echo ">> Scale down after HA tests"
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=1
+}
+
+function kind_e2e_istio() {
+    export GATEWAY_OVERRIDE=""
+    export GATEWAY_NAMESPACE_OVERRIDE=""
+
+    cluster_info
+
+    go test -race -count=1 -short -timeout=20m -tags=e2e ./test/conformance \
+        --enable-alpha --enable-beta \
+        --skip-tests="${ISTIO_UNSUPPORTED_E2E_TESTS}" \
+        --ingressendpoint="${IPS[0]}" \
+        --ingressClass=gateway-api.ingress.networking.knative.dev \
+        --cluster-suffix="$CLUSTER_SUFFIX"
+
+    # Give the controller time to sync with the rest of the system components.
+    sleep 30
+
+    echo ">> Scale up controller for HA tests"
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=2
+
+    go test -count=1 -timeout=15m -failfast -parallel=1 -tags=e2e ./test/ha -spoofinterval="10ms" \
+        --enable-alpha --enable-beta \
+        --ingressendpoint="${IPS[0]}" \
+        --ingressClass=gateway-api.ingress.networking.knative.dev \
+        --cluster-suffix="${CLUSTER_SUFFIX}"
+
+    echo ">> Scale down after HA tests"
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=1
+}
```

Comment:
This feel similar-but-not-identical to both `kind_e2e_contour` and `e2e_istio`. Might be worth a TODO to deduplicate more _in a future PR_.